### ISSUE TITLE
mosh fails to start correctly after upgrade to Tailscale 1.26.1 + ssh

### ISSUE THREAD

**Issue by vielmetti** [tailscale/tailscale#4919]
### What is the issue?

After turning on "tailscale ssh" on a server, mosh fails to start with the error message

```
/opt/homebrew/bin/mosh: Did not find mosh server startup message. (Have you installed mosh on your server?)
```

### Steps to reproduce

Upgrade Tailscale on a server to v1.26.1, turn on "tailscale ssh", try to connect to server with "mosh user@tailscale-node"

### Are there any recent changes that introduced the issue?

It looks like #4752 addresses this very issue, which is one thing that confuses me - I *think* that 1.26.1 has this patch in it.

### OS

Linux, macOS

### OS version

Debian 10 (buster)

### Tailscale version

1.26.1

### Bug report

BUG-739f04a08adef6a9fb63d380f766a3fd2ab3e4d7a2c586f9b21e720db5b4fe72-20220624030033Z-34749df6e9fcfebd

**Comment by DentonGentry** [comment]
We do have people successfully using mosh, though we also have a couple people who run into a similar issue with mosh-server not being available.

Does `ssh -T servername -- mosh-server --help` print a usage message from mosh-server? If so, that would indicate that $PATH works.

We had one report that `mosh --verbose servername` works but doesn't work without `--verbose`. We don't understand why yet, but if you see the same effect that would be interesting.

**Comment by vielmetti** [comment]
I was able to get a successful connection with `mosh --no-ssh-pty user@tailscale-host`, where `mosh user@tailscale-host` is failing.

Once I saw this error message

```
Bad MOSH CONNECT string: MOSH CONNECT 60001 3Fyuy71RtW45pFpfxmVBHgConnection to tailscale-host closed.
```

but it's not consistent.

My version of `mosh` is `mosh 1.3.2 [build mosh 1.3.2]` on Apple MacBook Air M1, installed with homebrew. It doesn't have a `mosh --verbose` option.

**Comment by apenwarr** [comment]
I just tested on two client devices (Blink Shell's mosh on iOS, and mosh 1.3.2 from homebrew on macOS) connecting to an ubuntu linux machine running tailscale 1.25.98 and then 1.27.57, both times with tailscale --ssh enabled. mosh on the server is also 1.3.2.

I didn't have to do anything special and it worked for me. just `mosh <hostname>` from both Blink and macOS.

**Comment by apenwarr** [comment]
Note: even though mosh works for me, when there is a mosh session, it prevents tailscaled from upgrading safely. See https://github.com/tailscale/tailscale/issues/4992

**Comment by apenwarr** [comment]
For completeness, I also did a localhost mosh (over tailscale) from my linux box to my linux box, and that worked too.

**Comment by spraza** [comment]
> I didn't have to do anything special and it worked for me. just `mosh <hostname>` from both Blink and macOS.


I was trying `mosh username@100.x.x.x` and was running into the error in the issue description. But `mosh 100.x.x.x` worked fine, and afterwards `mosh username@100.x.x.x` also worked. Not sure why this is the case, but thought to share here in case it helps root causing the problem.

**Comment by alock** [comment]
So far I can only get mosh and tailscale ssh (which I run in accept mode) to work by using `mosh --no-ssh-pty`.  I can successfully use ssh to the same "server" which is just a raspberrypi.  I installed mosh/mosh-server on the raspberrypi with basic `apt get mosh`.

Failing Clients:
* macbook with mosh 1.3.2 installed via homebrew (error messages below)
`/usr/local/bin/mosh: Did not find mosh server startup message. (Have you installed mosh on your server?)`
* ipad with Blink Shell app v15.0.8.542 (error messages below)
`Incorrect mosh-server startup sequence`

Both clients above successfully mosh into the raspberry pi with the `--no-ssh-pty` flag.

**Comment by btaylor84** [comment]
I'm having the same error when connecting from Windows WSL2 Ubuntu to an instance in the Oracle cloud.

using `mosh --no-ssh-pty` DOES work for me, though.

**Comment by DentonGentry** [comment]
Is the system you're SSHing *to* have Tailscale 1.26.2 or 1.26.1 installed?

**Comment by btaylor84** [comment]
yes, 1.26.2

**Comment by alock** [comment]
Updated to `tailscale 1.26.2`on the pi as well still same issue.  All ssh checks with the output below I believe are using tailscale ssh because I removed all `~/.ssh/authorized_keys`.

```shell
❯ ssh pi 'tailscale version'
1.26.2
  tailscale commit: 9e47908f1dcf48c567506be6ba9edc506072025a
  other commit: 27ee7cf2f2be732769c12494276af9b8cbbd0ca7
  go version: go1.18.3-ts04d67b90d8
❯ ssh pi 'echo $PATH'
/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/sbin:.
❯ ssh pi 'which mosh'
/usr/bin/mosh
❯ mosh pi
Connection to {tailnet.fqdn} closed.
/usr/local/bin/mosh: Did not find mosh server startup message. (Have you installed mosh on your server?)
```

**Comment by bradfitz** [comment]
I can repro the failure.

From macOS (homebrew mosh 1.3.2) to Linux Tailscale SSH mosh 1.3.2, using tailscaled at main as of Friday July 8th (d6b8a18b09d2c24111777d51f79fd63680023438).

Works with `mosh --no-ssh-pty`, but in PTY mode it's sporadic. It usually fails but works every few tries. Sometimes every 3 or 4 tries. Sometimes after 10 tries.

tailscaled logs says:

```
ssh-session(sess-20220711T032257-1b468504ce): starting pty command: [/home/bradfitz/go/bin/tailscaled be-child ssh --uid=1001 --gid=1001 --groups=1001,27,105,115 --local-user=bradfitz --remote-user=bradfitz@tailscale.com --remote-ip=100.120.208.62 --has-tty=true --tty-name=pts/25 --cmd=/bin/bash -- -c mosh-server 'new' '-c' '256' '-s' '-l' 'LANG=en_US.UTF-8']
ssh-session(sess-20220711T032257-1b468504ce): stdout copy: read /dev/ptmx: file already closed
ssh-session(sess-20220711T032257-1b468504ce): terminating SSH session from 100.120.208.62: read /dev/ptmx: file already closed
ssh-session(sess-20220711T032257-1b468504ce): Wait: code=-1
```

So I suspect it's a stdout/stderr close order/dup/race issue.

/cc @maisem

**Comment by bradfitz** [comment]
Ah, I think it has something to do with how `mosh-server` daemonizes itself.

When I start a mosh session, it runs remotely:

```
mosh-server new -c 256 -s -l LANG=en_US.UTF-8
```

If I strace that, it ends in:

```
write(1, "\r\n", 2
)                     = 2
write(1, "\n", 1
)                       = 1
getsockname(4, {sa_family=AF_INET, sin_port=htons(60002), sin_addr=inet_addr("100.74.70.3")}, [128->16]) = 0
write(1, "MOSH CONNECT 60002 xxxxxxxx"..., 42MOSH CONNECT 60002 xxxxxxxx
) = 42
rt_sigaction(SIGHUP, {sa_handler=SIG_IGN, sa_mask=~[RTMIN RT_1], sa_flags=SA_RESTORER, sa_restorer=0x7f74495f0d60}, NULL, 8) = 0
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=~[RTMIN RT_1], sa_flags=SA_RESTORER, sa_restorer=0x7f74495f0d60}, NULL, 8) = 0
clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f7449589e50) = 1994273
write(2, "\nmosh-server (mosh 1.3.2) [build"..., 45
mosh-server (mosh 1.3.2) [build mosh 1.3.2]
) = 45
write(2, "Copyright 2012 Keith Winstein <m"..., 51Copyright 2012 Keith Winstein <mosh-devel@mit.edu>
) = 51
write(2, "License GPLv3+: GNU GPL version "..., 201License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

) = 201
write(2, "[mosh-server detached, pid = 199"..., 38[mosh-server detached, pid = 1994273]
) = 38
ioctl(1, TCGETS, {B38400 opost isig icanon echo ...}) = 0
ioctl(1, TCSBRK, 1)                     = 0
ioctl(2, TCGETS, {B38400 opost isig icanon echo ...}) = 0
ioctl(2, TCSBRK, 1)                     = 0
exit_group(0)                           = ?
+++ exited with 0 +++
```

Note the `TCSBRK` on stdout & stderr with the "1" arg (non-zero).

https://man7.org/linux/man-pages/man4/tty_ioctl.4.html ...

```
   Sending a break
       TCSBRK Argument: int arg

              Equivalent to tcsendbreak(fd, arg).
...
              (SVr4, UnixWare, Solaris, and Linux treat
              tcsendbreak(fd,arg) with nonzero arg like tcdrain(fd).
```

https://pubs.opengroup.org/onlinepubs/009696799/functions/tcdrain.html

> The tcdrain() function shall block until all output written to the object referred to by fildes is transmitted. The fildes argument is an open file descriptor associated with a terminal.
>
> Any attempts to use tcdrain() from a process which is a member of a background process group on a fildes associated with its controlling terminal, shall cause the process group to be sent a SIGTTOU signal. If the calling process is blocking or ignoring SIGTTOU signals, the process shall be allowed to perform the operation, and no signal is sent.

And indeed, tailscaled logs:

```
ssh-session(sess-20220711T142619-96945e2098): wait: signal: hangup
ssh-session(sess-20220711T142619-96945e2098): Wait: code=-1
```

**Comment by bradfitz** [comment]
With recording on, I see the difference between a failure and success, respectively:

```
bradfitz@tsdev:~/src/tailscale.com$ sudo less /var/lib/tailscale/ssh-sessions/ssh-session-1657552565774527760-4140888141.cast
{"version":2,"width":185,"height":68,"timestamp":1657552565,"env":{"TERM":"xterm-256color"}}
[0.02469105,"o","\r\r\n\r\n"]

bradfitz@tsdev:~/src/tailscale.com$ sudo less /var/lib/tailscale/ssh-sessions/ssh-session-1657552778649140780-1418332916.cast
{"version":2,"width":185,"height":68,"timestamp":1657552778,"env":{"TERM":"xterm-256color"}}
[0.020382635,"o","\r\r\n\r\nMOSH CONNECT 60004 Z2YXHyFOcxxxxxg1YzNw\r\n"]
```

Debugging scribbles are in https://github.com/tailscale/tailscale/commit/31eb06f020020f53548248655d98318cedc986b9

**Comment by asineth0** [comment]
Same issue using `mosh` between two hosts w/ Tailscale SSH. Using `mosh --no-ssh-pty` works, without the additional argument it is unreliable and doesn't work consistently.

**Comment by maisem** [comment]
@asineth0 what version are you running? Can you provide a `tailscale bugreport` id?

**Comment by bradfitz** [comment]
This should've been fixed by af412e8874e94dc3ac57c37c3ec5e0606aa08fbb in 1.28.0.

Please file a new bug report against 1.28.0 if it's not fixed and we'll take a look.

**Comment by vielmetti** [comment]
If you ran into this bug in 1.26.1 with mosh, you might well also run into #5029 or #5095 when upgrading from 1.26.1 to 1.28.0.

**Comment by btaylor84** [comment]
yes, 1.26.2

**Comment by DentonGentry** [comment]
At this point as we've fixed other SSH-related bugs since 1.26.x, please install 1.30.2.
We have an additional related change coming in 1.32, in https://github.com/tailscale/tailscale/pull/5866
